# Applied Statistics - Assessment Problems
***

## Problem 1: Extending the Lady Tasting Tea

Let's extend the Lady Tasting Tea experiment as follows. The original experiment has 8 cups: 4 tea-first and 4 milk-first. Suppose we prepare 12 cups: 8 tea-first and 4 milk-first. A participant claims they can tell which was poured first.

Simulate this experiment using ```numpy``` by randomly shuffling the cups many times and calculating the probability of the participant correctly identifying all cups by chance. Compare your result with the original 8-cup experiment.

In your notebook, explain your simulation process clearly, report and interpret the estimated probability, and discuss whether, based on this probability, you would consider extending or relaxing the p-value threshold compared to the original design

In [22]:
# Importing the required libraries.
# Mathematical functions from the standard library
import math 
# Permutations and combinations.
import itertools
# Random selection.
import random
# Numerical structures and operations.
import numpy as np 
# Plotting.
import matplotlib as plt 

### Background: The Original Lady Tasting Tea

In Fisher’s classic experiment:
- Total cups: 8  
- 4 cups poured **tea first**
- 4 cups poured **milk first**

The lady claimed she could correctly identify which was which.

### Simulation Setup: Extended Experiment

We now have:
- **12 cups total**
- **8 tea-first**
- **4 milk-first**

We’ll calculate the number of possible combinations of selecting 4 milk-first cups out of 12 and determine the probability of guessing all correctly by chance.

#### Cups of Tea

In [23]:
# Number of cups of tea in total.
no_cups = 12

# Number of cups of tea with tea in first.
no_cups_tea_first = 8

# Number of cups of tea with milk in first.
no_cups_milk_first = 4

#### Combinations using `math.comb(n, k)`

The function `math.comb(n, k)` calculates the number of ways (combinations) of choosing `k` elements from a set of `n` elements,  
where the order of selection does not matter and no element is chosen more than once.

- We have to be aware that both `n` and `k` must be _non-negative integers_ (whole numbers).  
- This means they can include _0_, so they are not restricted to natural numbers (which start from 1).  
- The function will raise a 'ValueError' if either `n` or `k` is negative or not an integer.  
- Additionally, `k` cannot be greater than `n`.

In our case, we use it to calculate the number of ways to select 4 milk-first cups from a total of 12 cups.

In [24]:
# Number of combinations of selecting four cups from 12.
no_comb = math.comb(no_cups, no_cups_milk_first)

# Show combinations.
no_comb

495

Based on the result, the participant (the “lady tasting tea”) has **1 in 495** chance of randomly getting all cups correct.

We can also verify this by manually computing the number of combinations.

In [25]:
# Manual calculation using ordered selection

# Number of ways to choose 4 cups out of 12 (order matters)
ways_ordered = 12 * 11 * 10 * 9
print("Number of ways to select 4 cups out of 12 (ordered):", ways_ordered)

# Number of ways to shuffle those 4 chosen cups
no_shuffles = 4 * 3 * 2 * 1
print("Number of ways to shuffle 4 cups:", no_shuffles)

# Number of combinations (order removed)
no_comb = ways_ordered // no_shuffles
print("Number of combinations:", no_comb)


Number of ways to select 4 cups out of 12 (ordered): 11880
Number of ways to shuffle 4 cups: 24
Number of combinations: 495


#### How Does `math.comb(n, k)` Work?

The formula for `math.comb(n, k)` is:

$\binom{n}{k} = \frac{n!}{k!(n-k)!}$.

In our experiment:
- \( n = 12 \): total number of cups  
- \( k = 4 \): number of cups poured milk first

Breaking down the calculation:
- \( n! = 12! \)
- \( k! = 4! \)
- \( (n-k)! = (12 - 4)! = 8! \)

Substituting the values:

$\binom{12}{4} = \frac{12!}{4!(8)!} = 495 $

In [26]:
# Calculating the combination using the factorial formula from math.comb function.
n = 12
k = 4

math.factorial(n) // (math.factorial(k) * math.factorial(n - k))


495